### PR TITLE
soc: arm: atmel_sam: sam4: Enable MPU/FPU

### LINKED ISSUE #23947: soc: arm: atmel: sam4e: Enable FPU
**Is your enhancement proposal related to a problem? Please describe.**
Enable FPU support

**Additional context**
Atmel SAM4E have a Cortex-M4F FPU that implements the FPv4-SP floating-point extension.
The FPU fully supports single-precision add, subtract, multiply, divide, multiply and accumulate, and square root operations. It also provides conversions between fixed-point and floating-point data formats, and floating-point constant instructions.
The FPU provides floating-point computation functionality that is compliant with the ANSI/IEEE Std 754-2008, IEEE Standard for Binary Floating-Point Arithmetic, referred to as the IEEE 754 standard.
The FPU contains 32 single-precision extension registers, which can also be accessed as 16 doubleword registers for load, store, and move operations.

### PR DESCRIPTION
**Pull request by nandojve**
Add missing definitions to enable MPU and FPU support when appropriated.

fixes #23947
~includes the commits from #24487 (blocked by #24487)~

### CODE REVIEW

File: soc/arm/atmel_sam/sam4e/Kconfig.series
Comment (stephanosio) [inline soc/arm/atmel_sam/sam4e/Kconfig.series:11]: `CPU_HAS_ARM_MPU` is the symbol that indicates that ARM MPU is available on an SoC, and`ARM_MPU` is the symbol used to enable it.

Unless it is a requirement that you enable the ARM MPU for this SoC, `ARM_MPU` should be selected by the board using this SoC.

File: soc/arm/atmel_sam/sam4e/Kconfig.series
Comment (stephanosio) [inline soc/arm/atmel_sam/sam4e/Kconfig.series:13]: Until the ARM FP symbols are refactored, for now, this can be thought of as implied when `CPU_HAS_FPU` is selected for the Cortex-M4 and Cortex-M7 (ARMv7E-M).

File: soc/arm/atmel_sam/sam4s/Kconfig.series
Comment (stephanosio) [inline soc/arm/atmel_sam/sam4s/Kconfig.series:11]: same here

File: soc/arm/atmel_sam/sam4e/Kconfig.series
Comment (nandojve) [reply soc/arm/atmel_sam/sam4e/Kconfig.series]: Done!

File: soc/arm/atmel_sam/sam4e/Kconfig.series
Comment (nandojve) [reply soc/arm/atmel_sam/sam4e/Kconfig.series]: Done!

File: soc/arm/atmel_sam/sam4s/Kconfig.series
Comment (nandojve) [reply soc/arm/atmel_sam/sam4s/Kconfig.series]: Done!

File: soc/arm/atmel_sam/sam4e/soc.h
Comment (stephanosio) [inline soc/arm/atmel_sam/sam4e/soc.h:44]: I suppose you are including `devicetree.h` here (inside `#ifndef _ASMLANGUAGE` because of the DT symbol references below.
```
/** Processor Clock (HCLK) Frequency */
#define SOC_ATMEL_SAM_HCLK_FREQ_HZ DT_ARM_CORTEX_M4F_0_CLOCK_FREQUENCY
/** Master Clock (MCK) Frequency */
#define SOC_ATMEL_SAM_MCK_FREQ_HZ SOC_ATMEL_SAM_HCLK_FREQ_HZ
```

In that case, one of the following should be done:
1. Move the defs into `#ifndef _ASMLANGUAGE` if these defs need not be referenced in ASM code OR
2. Move the inc to outside `#ifndef _ASMLANGUAGE` so that these defs are also valid in ASM code.

File: soc/arm/atmel_sam/sam4s/soc.h
Comment (stephanosio) [inline soc/arm/atmel_sam/sam4s/soc.h:56]: same issue here

File: soc/arm/atmel_sam/sam4e/soc.h
Comment (nandojve) [reply soc/arm/atmel_sam/sam4e/soc.h:44]: I'm follow same approach found on SAME70/SAMV71 but is clear that I need fix that too :D

**Comment by nandojve** [comment]
West got merged so removing DNM.

**Review by stephanosio — APPROVED** [review]

File: dts/arm/atmel/sam4e.dtsi
Comment (ioannisg) [inline dts/arm/atmel/sam4e.dtsi:27]: This entry is not really needed - Cortex-M4 has a fixed number of 8 regions.

File: dts/arm/atmel/sam4s.dtsi
Comment (ioannisg) [inline dts/arm/atmel/sam4s.dtsi:28]: same here, the MPU node does not seem needed.

**Review by ioannisg — COMMENTED** [review]
Looks good - 
I 've got a comment about the MPU DT entries

File: dts/arm/atmel/sam4e.dtsi
Comment (nandojve) [reply dts/arm/atmel/sam4e.dtsi:27]: Hi @ioannisg, I prefer keep the entries, it could be important on future Zephyr versions. On my perspective it should be dropped only when it wont be necessary for all SoCs.

Comment (stephanosio) [reply dts/arm/atmel/sam4e.dtsi:27]: I agree with @nandojve here. Most (if not all) of the MPU-equipped SoCs seem to have the `mpu` entry and we should add one here for the sake of consistency; it is generally a good practice to retrieve hardware parameters from the device tree too, instead of hard-coding in the code.

Comment (ioannisg) [reply dts/arm/atmel/sam4e.dtsi:27]: No @stephanosio that cannot be accurate. AFAIK, only Cortex-M7 and Cortex-M33 SoCs have the mpu node defined.

Comment (stephanosio) [reply dts/arm/atmel/sam4e.dtsi:27]: @ioannisg You are right; that does seem to be the case:
https://github.com/zephyrproject-rtos/zephyr/search?p=1&q=extension%3Adtsi+e000ed90&unscoped_q=extension%3Adtsi+e000ed90

I wonder if we should add all the missing `mpu` entries for the all the MPU-equipped SoCs. cc @galak

Comment (ioannisg) [reply dts/arm/atmel/sam4e.dtsi:27]: We can - I was thinking this could eventually lead to having MPU enabled entirely.
But it is not required for now and most M3,M4 SoCs don't do that.

Comment (stephanosio) [reply dts/arm/atmel/sam4e.dtsi:27]: well, as long as we set `status = "disabled";` for the MPU, the MPU should be considered to be disabled (and handled as such in the code).

sure, adding MPU entries for the rest would be beyond the scope of this PR. maybe we should remove these entries until we do a refactoring in the future -- no strong opinion on this one.

Comment (ioannisg) [reply dts/arm/atmel/sam4e.dtsi:27]: agreed

Comment (nandojve) [reply dts/arm/atmel/sam4e.dtsi:27]: I think @galak already started changes on #24386. This way, should we wait?

Comment (ioannisg) [reply dts/arm/atmel/sam4e.dtsi:27]: no, I think #24386 does not add mpu nodes in dts for existing SoCs - this work is orthogonal to that one.

Comment (ioannisg) [reply dts/arm/atmel/sam4e.dtsi:27]: But we could wait, since #24386 would potentially remove the fixup entries for mpu and nvic

**Comment by nandojve** [comment]
#24386 merged.

**Comment by nandojve** [comment]
rebase on top of #24487.
Added a proposal to change codeowners to nandojve get access to all atmel DTS.

File: CODEOWNERS
Comment (ioannisg) [inline CODEOWNERS:242]: This seems wrong - the last pattern match has preference so this will supersede @benpicco 

Honestly, I don't think there's so much need for clean up here - CODEOWNERS file is meant to provide fine granularity in assigning reviews.

File: soc/arm/atmel_sam/sam3x/soc.h
Comment (ioannisg) [inline soc/arm/atmel_sam/sam3x/soc.h:39]: I think you should bring back the inclusion of devicetree.h here.

File: soc/arm/atmel_sam/sam4e/dts_fixup.h
Comment (ioannisg) [inline soc/arm/atmel_sam/sam4e/dts_fixup.h:2]: I guess this file will disappear but for now it's ok to add copyight

**Review by ioannisg — CHANGES_REQUESTED** [review]
thanks for this PR
Got some minor change requests, see my comments

**Comment by stephanosio** [comment]
Modified PR description since #24487 is not closed and still actively under review (hence, this PR does not "supersede" #24487):
```
-supersedes #24487
+includes the commits from #24487 (blocked by #24487)
```

**Comment by nandojve** [comment]
Rebase and drop CODEOWNERS proposal.

**Comment by galak** [comment]
LGTM, but will wait for @ioannisg approval.

**Review by ioannisg — APPROVED** [review]

**Pull request merged**
